### ISSUE TITLE
"Scratch space" under buck-out for end-user use?

### ISSUE THREAD

**Issue by thoughtpolice** [facebook/buck2#231]
Generally speaking I hate to pollute the top level directory with build artifacts, but every so often this is unavoidable for various reasons. On the other hand, I don't want to put things in `buck-out` if it might confuse buck because things appeared there. For example a shell script I have might download something temporarily and it's not invoked by `buck run`.

As an example, I have a script that starts buildbarn in a virtual machine, and requires me to create a "virtual" hard disk for the virtual machine to use. I kind of don't know where to put it, but `buck-out/` is tantalizing since it's already in my watchman ignore, sapling ignore, etc. (Making the shell script to start this a `buck run` target is probably possible even with the "loopy" nature of then needing to restart the daemon...)

Would it be possible or is it possible to have a a "scratch space" or "cellar" underneath `buck-out/v2` or something where you can just put your own temporary files? I.e. `buck-out/v2/cellar` or `buck-out/v2/scratch`, and the only guarantee is basically that 1) you can create this directory if you want and 2) buck will not touch it and so 3) you're responsible for cleaning it up.

(If this comes across as too odd or niche of a request I'm happy to take a "No", of course.)

**Comment by stepancheg** [comment]
Why not create another top-level directory in the repo, and mark it as `project.ignore` and add it to `.gitignore`?

**Comment by thoughtpolice** [comment]
Because `buck-out/` is already there, already in my ignore lists &mdash; it's tidier that way :]

It's a pretty marginal request, I admit. And actually, maybe this is also an "X/Y problem" sort of situation, but: if it was configurable, my preferred solution would be to actually change the path of `buck-out` itself to be nested inside another directory that *I* had control of i.e. I would change it to `output/buck/` or something instead and then I could do whatever I wanted with the rest of `output/`. But that doesn't seem possible to configure, and maybe shouldn't be, either (i.e. probably a lot of plumbing for marginal gain.) So at least having a carve-out under `buck-out/v2` would be just as good, for me.

**Comment by ndmitchell** [comment]
The problem with changing `buck-out` is that it's easy to implicitly depend on details about it in rules (e.g. it is 3 components up from something), so having that variability makes things easy to break.

The problem with having `buck-out/v2/scratch` be for other people is that then we have questions about lifetime, who cleans up there etc. In practice, people do tend to shove stuff in `buck-out` (for all the reasons you list) and we say "that's not supported" but don't break it.

**Comment by thoughtpolice** [comment]
Wouldn't issue 1 be partially mitigated through hermetic builds, though? Like if you declared `../../file.txt` as an input or whatever, relying on this behavior &mdash; you're sort of asking for what you get, right? If that file isn't committed in the repo or created by some other rule, and it doesn't exist, the build will of course fail, but I think that's a case with any "dangling leaf" in the graph, I think, right? You mis-specified your inputs. It could just as well refer to `some-other-file-that-doesnt-exist.txt` too. But it's more opaque and less easy to understand in the first case, though.

Side note but, I think another related reason why people do/want stuff like this and why people like like "1 directory containing everything" style designs &mdash; at least in my case &mdash; is because it means you can literally just `rm -rf $OUTDIR` in order to clean *everything*. It sounds trivial but I really do rely on this, I do it all the time. It's obvious, works everywhere, and you don't need to run any special commands for this to work contrast e.g. `make clean` which can actually *cause* rebuilding to occur before deleting files! And it may not delete everything. Or `git clean -xfd` to delete artifacts, which can actually destroy your working copy if you haven't used `git add` recently, and doesn't work in e.g. Sapling.

I honestly think the lifetime issue about a `scratch/` dir is one of those things where it's like, if you break it &mdash; you get to keep both pieces, that's on you. With reliable hermetic execution I think it's also generally less of a big deal. You could also add some various special cases e.g. "No relative source artifact may begin with `buck-out/v2/scratch`" or whatever. But I can understand how philosophically it's rather against the grain, where Buck genuinely wants you to do things its way (for good reason) and so it outright bans incorrect things, too.

**Comment by krallin** [comment]
>  it means you can literally just rm -rf $OUTDIR in order to clean everything. It sounds trivial but I really do rely on this, I do it all the time. It's obvious, works everywhere, and you don't need to run any special commands for this to work

FWIW, doing this with Buck2 will in fact tend to make Buck2 rather unhappy with you since you just deleted a bunch of files Buck2 believes are on disk

**Comment by thoughtpolice** [comment]
Yeah, and sometimes I remember to `buck clean` instead since it kills the daemon first (I used to always just `git clean -xfd`). But "delete the build entirely and restart" is so common and universal that it has strong mental inertia. People will want to do this instinctively. Mind you that doesn't mean you have to *support it* or change anything. I'm just explaining the mental background.
